apps/server/src/shared/repo/base.do.repo.ts

```diff
@@ -10,45 +10,96 @@ export type EntityProperties<P> = P & IBaseEntityProps;
+
+	/**
+	 * Maps a database entity to a domain object.
+	 * @param entity The database entity
+	 * @returns The domain object
+	 * @protected
+	 */
 	protected abstract mapEntityToDO(entity: E): T;
 
-	protected mapDOToEntityWithId(edo: T): EntityProperties<P> {
-		const entity: EntityProperties<P> = this.mapDOToEntity(edo);
-		entity.id = edo.id;
-		return entity;
+	/**
+	 * Maps a domain object to its related entity properties.
+	 * @param entityDO The domain object
+	 * @returns The mapped entity properties with id
+	 * @protected
+	 */
+	protected mapDOToEntityWithId(entityDO: T): EntityProperties<P> {
+		const entityProps: EntityProperties<P> = this.mapDOToEntity(entityDO);
+		entityProps.id = entityDO.id;
+		return entityProps;
 	}
 
+	/**
+	 * Uses the database entity constructor to create a new entity.
+	 * @param Type The constructor method for the entity
+	 * @param props The entity properties that are used to call the constructor
+	 * @returns A newly constructed database entity
+	 * @protected
+	 */
 	protected entityFactory(Type: new (I) => E, props: P): E {
 		return new Type(props);
 	}
 
-	async save(entityDos: T | T[]): Promise<T | T[]> {
-		const isArray = Array.isArray(entityDos);
-		const dos: T[] = isArray ? entityDos : [entityDos];
+	async save(entityDOs: T): Promise<T>;
+	async save(entityDOs: T[]): Promise<T[]>;
+	async save(entityDOs: T | T[]): Promise<T | T[]>;
+
+	/**
+	 * Saves one or more domain objects to the database.
+	 * @param entityDOs The domain objects to save
+	 * @returns The saved domain objects
+	 */
+	async save(entityDOs: T | T[]): Promise<T | T[]> {
+		const isArray = Array.isArray(entityDOs);
+		const dos: T[] = isArray ? entityDOs : [entityDOs];
 
 		const entities: E[] = await Promise.all(
-			dos.map(async (d) => {
-				const entityProps: EntityProperties<P> = this.mapDOToEntityWithId(d);
+			dos.map(async (domainObject: T): Promise<E> => {
+				const entityProps: EntityProperties<P> = this.mapDOToEntityWithId(domainObject);
 				const newEntity: E = this.entityFactory(this.getConstructor(), entityProps);
 
-				const entity: E = await this._em
-					.findOneOrFail(this.entityName, d.id as FilterQuery<E>)
-					.then((fetchedEntity: E) => {
-						// Ignore base entity properties when updating entity
-						const wrapped: EntityDTO<E> = wrap(newEntity).toObject(baseEntityProperties);
-
-						this.logger.debug(`Update entity with id ${fetchedEntity.id}`);
-						return Object.assign(fetchedEntity, wrapped) as E;
-					})
-					.catch(() => {
-						this.logger.debug(`Created new entity`);
-						return this._em.create(this.entityName, newEntity);
-					});
+				let entity: E;
+
+				if (!domainObject.id) {
+					// Create
+					entity = this._em.create(this.entityName, newEntity);
+					this.logger.debug(`Created new entity with id ${entity.id}`);
+				} else {
+					// Update
+					entity = await this._em
+						.findOneOrFail(this.entityName, domainObject.id as FilterQuery<E>)
```

Comment:
I changed it to `{ id: domainObject.id } as FilterQuery<E>` to make it more readable. It needs the cast, because the compiler can not fully resolve the id field of E in the FilterQuery for some reason. Let me know if you know why.